### ISSUE TITLE
Gradle daemons should shut down more aggressively

### ISSUE THREAD

**Issue by ZacSweers** [gradle/gradle#14741]
### Expected Behavior

I'm not sure how exactly these _should_ behave, but I believe the current situation is bad and should be improved.

### Current Behavior

If I switch around 2-3 projects, I quickly amass a collection of memory-intensive Gradle daemons that appear to linger indefinitely. After a certain point, these + the kotlin daemons kick my computer's fans into high gear until I go into activity monitor and start killing some off them off. I think Gradle should try to be more mindful of memory usage by default, possibly detecting unused daemons or offering some degree of "shut down daemons if they haven't run after a certain time period".

My computer is currently crawled to a stop with this

![image](https://user-images.githubusercontent.com/1361086/94985274-310f1b80-0523-11eb-886b-783304e14955.png)

```
$ jps|grep -E 'KotlinCompileDaemon|GradleDaemon'
9090 KotlinCompileDaemon
19753 GradleDaemon
25360 KotlinCompileDaemon
19762 KotlinCompileDaemon
16406 GradleDaemon
9083 GradleDaemon
20155 GradleDaemon
17212 GradleDaemon
24381 GradleDaemon
```

In short, I don't think I should have to go in and start manually killing off these processes myself.

### Context
I work a lot in OSS projects and regularly bounce around different projects and branches throughout the day. This is a common case.

### Steps to Reproduce 
Build 3-4 different projects to quickly reproduce this. Bonus points if any use Kotlin

### Your Environment
Build scan URL: https://scans.gradle.com/s/3cs4ev6q3fp72

**Comment by ZacSweers** [comment]
(to be clear - there might be settings for this that I'm not aware of, but in that case I'm also arguing that the defaults seem too lenient)

**Comment by runningcode** [comment]
How long were these daemons alive and inactive? 
https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:how_can_i_stop_a_daemon

**Comment by ZacSweers** [comment]
I don't know how to check that but somewhere in the realm of a couple hours

**Comment by runningcode** [comment]
You can configure the timeout to whatever idle time you like using the `org.gradle.daemon.idletimeout` property.
https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties

**Comment by big-guy** [comment]
Are these different versions of Gradle or just incompatible (due to memory or other JVM properties)?

**Comment by ZacSweers** [comment]
I suppose, but it's not always clear to me why daemons aren't reused. These are multiple projects potentially using different gradle or java versions

**Comment by ZacSweers** [comment]
I see from that doc that the timeout is 3 hours by default. That seems quite high, and I'd be curious what the motivation is. I feel like 30 minutes or less would be better? Someone taking a break from builds for that long probably would rather keep the battery and pay the daemon startup cost when coming back.

**Comment by jonathanlermitage** [comment]
A new gradle daemon is started if gradle version, jvm version or jvm parameters changes (even when you specify a Xms, a -Dfile.encoding...). Maybe the gradle daemon could be a bit smarter, I don't know.  At least, it could specify a very low Xms (like Xms32m):  sometimes, you see that your gradle daemon consumes 512MB or 1GB of Heap memory at startup, but you can reduce this consumption to ~100MB if you run a full GC manually (e.g. via VisualVM), and it won't grow when idle. 

Some tips:
- you can reduce the number of daemon by using the same version of gradle (I know, it may be impossible if you don't own the projects you contribute to ^_^).
- if you use an IDE like IntelliJ, you may want to use your local version of Gradle, not the gradle wrapper. This way, you will use the same version of gradle in every project and, maybe, don't start aditional daemons.
- update your IDE. IJ is designed to stop gradle daemons when you close your IDE (not projects), but some releases were bogus, they did not kill the gradle daemons.
- configure the gradle daemons globally, via a config file in your ~/.gradle/gradle.properties. You can force the Xms and Xmx flags, and reduce the 3h timeout. (a part of my) My gradle.properties:  
  ```
  org.gradle.daemon.idletimeout=3600000   <- 1h timeout (value is ms)
  org.gradle.jvmargs=-Xms256m -Xmx2g -XX:MaxMetaspaceSize=512m -Dfile.encoding=UTF-8
  ```

**Comment by jonathanlermitage** [comment]
Also, gradle and kotlin daemons are different things. If you see too many Kotlin compile daemon, even for non-Kotlin projects (I see kotlin daemons for some Java Maven projects too... ^_^), you should open an issue here: https://youtrack.jetbrains.com. I see you work with Android Studio, which is based on IntelliJ. These kotlin compile daemons are started by Android Studio, not by gradle.

**Comment by matejdro** [comment]
I'm also in a similar position when juggling different projects.

* Projects are shared with different people, so we are forced to use gradle wrapper (that way everyone uses same gradle version). For the same reason we are sometimes forced to use different versions of Gradle (project A updates, but project B did not update yet)
* Does Android Studio also auto-close daemons? I do not remember that ever happening, but maybe it were buggy releases you mentioned.
* Some of the projects are pretty large so I cannot reduce `Xmx`
* I explained the timeout issue here: https://github.com/gradle/gradle/issues/17890

Maybe solution could be a property that would instruct gradle to stop all idle incompatible daemons whenever new daemon starts? This would ensure that opening a new project would always wipe the slate clean of all previous daemons.

**Comment by ZacSweers** [comment]
Closing due to lack of feedback from Gradle

**Comment by martinbonnin** [comment]
For whoever comes here, this is what I have in my ~/.gradle/gradle.properties

```
# 5 min timeout for Gradle daemons
org.gradle.daemon.idletimeout=300000
```

Seems ok so far...
